engine-comparison/dispatcher.sh

```diff
@@ -189,21 +189,43 @@ measure_coverage () {
   rm -fr $CORPUS_DIR $SANCOV_DIR
   mkdir -p $CORPUS_DIR $SANCOV_DIR $REPORT_DIR
 
+  # Recall which trial
+  if [[ -f $REPORT_DIR/latest-trial ]]; then
+    . $REPORT_DIR/latest-trial
+  else
+    LATEST_TRIAL=0
+  fi
+
+  # Check for next trial
+  if [[ -d ${EXPERIMENT_DIR}/trial-$(($LATEST_TRIAL + 1)) ]]; then
+    # This round, we finish report for the old trial. But next time continue
+    echo "LATEST_TRIAL=$((LATEST_TRIAL + 1))" > $REPORT_DIR/latest-trial
+  else
+    echo "LATEST_TRIAL=$LATEST_TRIAL" > $REPORT_DIR/latest-trial
+  fi
+
+  # Enter trial directories
+  EXPERIMENT_DIR=${EXPERIMENT_DIR}/trial-${LATEST_TRIAL}
+  REPORT_DIR=${REPORT_DIR}/trial-${LATEST_TRIAL}
+  mkdir -p $REPORT_DIR
+
   # Use the corpus-archive directly succeeding the last one to be processed
-  if [[ -f $REPORT_DIR/latest-report ]]; then
-    . $REPORT_DIR/latest-report
+  if [[ -f $REPORT_DIR/latest-cycle ]]; then
+    . $REPORT_DIR/latest-cycle
   else
-    LATEST_REPORT=0
+    LATEST_CYCLE=0
   fi
 
-  THIS_CYCLE=$(($LATEST_REPORT + 1))
+
+  THIS_CYCLE=$(($LATEST_CYCLE + 1))
   while [[ $(grep "^${THIS_CYCLE}$" ${EXPERIMENT_DIR}/results/skipped-cycles) ]]; do
     THIS_CYCLE=$((THIS_CYCLE + 1))
   done
   if [[ ! -f ${EXPERIMENT_DIR}/corpus/corpus-archive-${THIS_CYCLE}.tar.gz ]]; then
     echo "On cycle $THIS_CYCLE, no new corpus found for benchmark $BENCHMARK and fengine $FENGINE_NAME"
     return
   fi
+  # TODO: should we skip these? bottleneck here could be fine with small wait period
```

Comment:
Does skipping cycles affect the graph?  We probably want all the data points in the graph.  But, at the same time we want relatively up-to-date reports.  Perhaps we could choose not to skip cycles and instead beef up the dispatcher VM as needed to generate timely reports.

Reply:
Regarding the graph, there's always the option to interpolate between empty values. But yes, these would be data points absent from the graph; this is also the case when the runner can't produce data because it's waiting for `gsutil`. 

Like you said, both of these can likely be solved by increasing cores. But even if the cores are powerful enough to almost never skip a cycle, is it still more prudent to account for the possibility?

Reply:
If we don't skip, we'll just be slightly behind in updating the webpage, but the data should still be correct.  I would argue that having all data points and being slightly behind in reporting the live graph is better than losing data points and being really fast at updating the graph.  And hopefully with enough cores we can have both.

Reply:
Yeah, this makes sense to me. I also realized that skipped cycles are easier to avoid in the dispatcher than the runner, so it's preferable to do so.

Reply:
> this is also the case when the runner can't produce data because it's waiting for gsutil

Does this block the runner from fuzzing?  Ideally we want every runner to fuzz constantly and not be blocked by bookkeeping operations.

Reply:
> Does this block the runner from fuzzing?

I think it depends on the relationship of `$JOBS` and cores, but in general yes. I've been thinking about, for example, having `JOBS=1` on a two core machine and running `gsutil` without `-m`; I think that this would totally partition jobs between cores and avoid fuzz disruption. `gsutil -m` is also fairly customizable, and could probably be configured to run on only one core.

Reply:
I think an approach like this would make for a better experimental framework and provide more accurate results.  But this is outside of the scope of this PR.  I'll open an issue.